apis/r/R/SOMADataFrame.R

```diff
@@ -327,9 +327,65 @@ SOMADataFrame <- R6::R6Class(
     #' @return None, instead a \code{\link{.NotYetImplemented}()} error is raised
     #'
     shape = function() stop(errorCondition(
-      "'SOMADataFrame$shape()' is not implemented yet",
+      "'SOMADataFrame$shape()' is not implemented",
       class = 'notYetImplementedError'
-    ))
+    )),
+
+    #' @description Retrieve the maxshape; as \code{SOMADataFrames} are shapeless,
+    #' simply raises an error
+    #'
+    #' @return None, instead a \code{\link{.NotYetImplemented}()} error is raised
+    #'
+    maxshape = function() stop(errorCondition(
+      "'SOMADataFrame$maxshape()' is not implemented",
+      class = 'notYetImplementedError'
+    )),
+
+    #' @description Returns the shape slot for the soma_joinid dim, if the array
+    #' has one.  This is an important test-point and dev-internal access-point.
+    #' (lifecycle: maturing)
+    #' @return An integer64 for the soma_joinid dim's shape, if soma_joinid is
+    #' a dim, else NA.
+    .maybe_soma_joinid_shape = function() {
+      retval <- as.integer64(maybe_soma_joinid_shape(
+        self$uri,
+        config=as.character(tiledb::config(self$tiledbsoma_ctx$context()))
+      ))
+      if (retval == as.integer64(-1)) {
+        NA
+      } else {
+        retval
+      }
+    },
+
+    #' @description Returns the maxshape slot for the soma_joinid dim, if the array
+    #' has one.  This is an important test-point and dev-internal access-point.
+    #' (lifecycle: maturing)
+    #' @return An integer64 for the soma_joinid dim's shape, if soma_joinid is
+    #' a dim, else NA.
+    .maybe_soma_joinid_maxshape = function() {
+      retval <- as.integer64(maybe_soma_joinid_maxshape(
+        self$uri,
+        config=as.character(tiledb::config(self$tiledbsoma_ctx$context()))
+      ))
+      if (retval == as.integer64(-1)) {
+        NA
+      } else {
+        retval
+      }
+    },
```

Comment:
These are only used for tests, and don't make use of internal object structures, so do we need these attached to `SOMADataFrame`? In the test files, we can (and probably should) just use

```R
sjid <- maybe_soma_joinid_shape(sdf$uri, config = as.character(tiledb::config(sdf$tiledbsoma_ctx$context()))
```

Tacking these on as public methods to the `SOMADataFrame` class (and only this class?) seems like overkill, especially for an API we may not want exposed or are using anywhere else

Also, I'm not a fan of the `maybe_` style function names. These aren't used anywhere else in tiledbsoma-R (or amongst R packages as a whole); instead the paradigm settled on for this package for accessors that may-or-may-not work is either
 - just return `NA` (like `SOMADataFrame$shape()` prior to #2531)
 - throw an error (like bounding boxes with `soma_array$used_shape()`)

Reply:
They're not used _yet_ but they will be deeper in-stack. See discussion on #2407.

They will be absolutely crucial in Python for experiment-level resize, and note that we do have append mode existing in Python so this is very pressing.

For R we do not have append mode (this is backlogged as #1630) and I debated whether to include these accessors in R or not. Reason to exclude: because not used yet. Reason to include: I don't want to add to the pile of parity work needing to be done by whoever implements #1630 someday -- hoping to raise the parity a bit here.

The prefix `maybe_` indicates that the return value may be `None` (in Python) or `NA` (in R), and it's a `std::optional<int64_t>` in C++. I'm happy to leave off the prefix for the R side if that sounds more R-cultural to you.

Reply:
> and don't make use of internal object structures

Nope -- these are wired all the way down to C++. All on this PR.

Reply:
> sjid <- maybe_soma_joinid_shape(sdf$uri, config = as.character(tiledb::config(sdf$tiledbsoma_ctx$context()))

I'm fine with this as a compromise between these two:

> For R we do not have append mode (this is backlogged as https://github.com/single-cell-data/TileDB-SOMA/issues/1630) and I debated whether to include these accessors in R or not. Reason to exclude: because not used yet. Reason to include: I don't want to add to the pile of parity work needing to be done by whoever implements https://github.com/single-cell-data/TileDB-SOMA/issues/1630 someday -- hoping to raise the parity a bit here.

Reply:
> They're not used yet but they will be deeper in-stack. See discussion on #2407.

I don't see where in #2407 these names are specified (maybe I'm missing something) or their behavior defined. I also think figuring out _how_ resize and append-mode will work in R (what their user-facing APIs are) would be important into deciding where these two accessors belong (public, private, public-but-dotted, external)

> The prefix maybe_ indicates that the return value may be [...]

I get that, but the R way would be no prefix and describing the "alternate" behavior in the docs (eg. returning `NA`/`NULL`/zero-length vector or raising an error)

Reply:
> I don't see where in https://github.com/single-cell-data/TileDB-SOMA/issues/2407 these names are specified (maybe I'm missing something) or their behavior defined

Sorry, I meant the PRs listed there.

Reply:
I'm still going to ask that the `maybe_` prefix be removed; either just dotted-methods or removed entirely with implementation to be deciced later (merge with `$shape()`/`$maxshape()`, move to private, or expose publicly-but-dotted)

Also, I noticed that these are methods in R, but properties in Python. Do we want this discrepancy between the two APIs or should we settle on both having these be methods or properties?

Reply:
@mojaveazure https://github.com/single-cell-data/TileDB-SOMA/pull/2953/commits/5312d49cfe08282110645c5faafcf6051951855e